api/admin/controller.py

```diff
@@ -1061,114 +1061,129 @@ class CustomListsController(CirculationManagerController):
+    def _create_or_update_list(self, library, name, entries, collections, id=None):
+        data_source = DataSource.lookup(self._db, DataSource.LIBRARY_STAFF)
 
-            if id:
-                is_new = False
-                list = get_one(self._db, CustomList, id=int(id), data_source=data_source)
-                if not list:
-                    return MISSING_CUSTOM_LIST
-                if list.library != library:
-                    return CANNOT_CHANGE_LIBRARY_FOR_CUSTOM_LIST
-                if old_list_with_name and old_list_with_name != list:
-                    return CUSTOM_LIST_NAME_ALREADY_IN_USE
-            elif old_list_with_name:
+        old_list_with_name = CustomList.find(self._db, data_source, name, library)
+
+        if id:
+            is_new = False
+            list = get_one(self._db, CustomList, id=int(id), data_source=data_source)
+            if not list:
+                return MISSING_CUSTOM_LIST
+            if list.library != library:
+                return CANNOT_CHANGE_LIBRARY_FOR_CUSTOM_LIST
+            if old_list_with_name and old_list_with_name != list:
                 return CUSTOM_LIST_NAME_ALREADY_IN_USE
-            else:
-                list, is_new = create(self._db, CustomList, name=name, data_source=data_source)
-                list.created = datetime.now()
-                list.library = library
+        elif old_list_with_name:
+            return CUSTOM_LIST_NAME_ALREADY_IN_USE
+        else:
+            list, is_new = create(self._db, CustomList, name=name, data_source=data_source)
+            list.created = datetime.now()
+            list.library = library
 
-            list.updated = datetime.now()
-            list.name = name
+        list.updated = datetime.now()
+        list.name = name
 
-            if entries:
-                entries = json.loads(entries)
-            else:
-                entries = []
-
-            old_entries = [x for x in list.entries if x.edition]
-            membership_change = False
-            for entry in entries:
-                pwid = entry.get("pwid")
-                work = self._db.query(
-                    Work
-                ).join(
-                    Edition, Edition.id==Work.presentation_edition_id
-                ).filter(
-                    Edition.permanent_work_id==pwid
-                ).one()
-
-                if work:
-                    entry, entry_is_new = list.add_entry(work, featured=True)
-                    if entry_is_new:
-                        membership_change = True
-
-            new_pwids = [entry.get("pwid") for entry in entries]
-            for entry in old_entries:
-                if entry.edition.permanent_work_id not in new_pwids:
-                    list.remove_entry(entry.edition)
+        if entries:
+            entries = json.loads(entries)
+        else:
+            entries = []
+
+        old_entries = [x for x in list.entries if x.edition]
```

Comment:
Here you're still going to take a performance hit if there are thousands of items in the list. You're going to pull all the  CustomListEntry objects from the CustomList, and then look up an individual Work for every entry in the new list. That's not a reason to not land this branch but I don't think it's going to make list membership _changes_ any faster.

You can look at `Collection.catalog_identifiers` and `CoverageRecord.bulk_add` for examples of how we deal with making this sort of membership change, when (say) a hundred records out of millions total are changing.

Reply:
That's a good point, and I think I'm going to need a follow-up branch to address performance on the custom list edit page anyway.